### Problem Statement:
Mobile carrier Megaline has found out that many of their subscribers use legacy plans.
They want to develop a model that would analyze subscribers' behavior and recommend
one of Megaline's newer plans: Smart or Ultra.


### Data analytic question
To develop a model that predicts the best mobile plan for a subscriber

### Measure of success

The accuracy of the model should be 0.75

In [9]:
# Importing our libraries
# ---
# import Pandas for data manipulation
import pandas as pd
#improt numpy for arithimetic oprations
import numpy as np
from sklearn.tree import DecisionTreeClassifier
# to split the data set import 
from sklearn.model_selection import train_test_split
# to calculate mse, mae and r2 import the function from sklearn
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [3]:
# Load and read the data as store it as plan_df data frame
plan_df = pd.read_csv("https://bit.ly/UsersBehaviourTelco")

### Since the data is clean as stated, we skip preparation and cleaning

## Solution

In [4]:
# split the data into train, valid and test
initial_train, test = train_test_split(plan_df, test_size = 0.2, random_state=40)
train, valid = train_test_split(initial_train, test_size = 0.1, random_state=40)

In [5]:
# checking the values in the plan column ['is_ultra']
train['is_ultra'].unique()

array([0, 1], dtype=int64)

### Observation
We have noted that the values in is_ultra are only 0 and 1 as expected

In [6]:
# preparing the train, valid, and test data sets
x_train= train.drop(['is_ultra'],  axis=1)
y_train = train['is_ultra']
x_test= test.drop(['is_ultra'],  axis=1)
y_test = test['is_ultra']
x_valid= valid.drop(['is_ultra'],  axis=1)
y_valid = valid['is_ultra']


#### Linear Regression

In [26]:
# Import linear regression from sklearn

from sklearn.linear_model import LinearRegression
lr = LinearRegression()

#building linear regression model
lrmodel = lr.fit(x_train, y_train)
#use model to predict
y_linear_pred = lrmodel.predict(x_test)

actual_l = y_linear_pred.mean()


# calculating the mse, mae 
linear_mae = mean_absolute_error(y_test,y_linear_pred)


linear_mse = mean_squared_error(y_test, y_linear_pred)

linear_rmse = mean_squared_error(y_test, y_linear_pred, squared= False)

print ('linear regressor mean :', actual_l, 'linear_mae: ', linear_mae,'linear_mse: ', linear_mse, 'linear rmse', linear_rmse) # for linear regression

linear regressor mean : 0.30266358736469995 linear_mae:  0.39755213238676046 linear_mse:  0.20364514353609944 linear rmse 0.45127058793599595


#### RandomForestRegressor

In [42]:
# import randomforestregressor from sk learn
from sklearn.ensemble import RandomForestRegressor

# Build and train RandomForestRegressor model
regressor = RandomForestRegressor(random_state=42)
regressor.fit(x_train, y_train)

# use the model to predict the mean 
y_regressor_pred = regressor.predict(x_test)
actual_rfr = y_regressor_pred.mean()

# calculating the mse, mae 

forest_mae = mean_absolute_error(y_test,y_regressor_pred)


forest_mse = mean_squared_error(y_test, y_regressor_pred)

forest_rmse = mean_squared_error(y_test, y_regressor_pred, squared= False)
print ('random forest mean :', actual_rfr, 'random_forest_mae: ', forest_mae,'random_forest mse: ', forest_mse ,'random_forest rmse: ', forest_rmse) # for RandomForestRegressor



random forest mean : 0.33110419906687405 random_forest_mae:  0.29287713841368584 random_forest mse:  0.1615710730948678 random_forest rmse:  0.40195904405158966


#### Decision tree

In [28]:
#import decision Tree model class fro sklearn
from sklearn.tree import DecisionTreeRegressor

#build and train the model

tree = DecisionTreeRegressor(random_state=12345)
tree.fit(x_train, y_train)

#calcualte the mean using the model
y_tree_pred = tree.predict(x_test)
actual_tree = y_tree_pred.mean()


#calculate the mae, mse

tree_mae = mean_absolute_error(y_test,y_tree_pred)


tree_mse = mean_squared_error(y_test, y_tree_pred)
tree_rmse = mean_squared_error(y_test, y_tree_pred, squared=False)
print ('DecisionTreeRegressor mean :', actual_tree, 'decision tree mae: ', tree_mae,'decison tree mse: ', tree_mse, 'decison tree rmse: ', tree_rmse) # for DecisionTreeRegressor


DecisionTreeRegressor mean : 0.3343701399688958 decision tree mae:  0.2986003110419907 decison tree mse:  0.2986003110419907 decison tree rmse:  0.5464433282985444


In [29]:
print ('linear regressor mean :', actual_l, 'linear_mae: ', linear_mae,'linear_mse: ', linear_mse, 'linear rmse', linear_rmse) # for linear regression
print ('random forest mean :', actual_rfr, 'random_forest_mae: ', forest_mae,'random_forest mse: ', forest_mse ,'random_forest rmse: ', forest_rmse) # for RandomForestRegressor
print ('DecisionTreeRegressor mean :', actual_tree, 'decision tree mae: ', tree_mae,'decison tree mse: ', tree_mse, 'decison tree rmse: ', tree_rmse) # for DecisionTreeRegressor


linear regressor mean : 0.30266358736469995 linear_mae:  0.39755213238676046 linear_mse:  0.20364514353609944 linear rmse 0.45127058793599595
random forest mean : 0.33110419906687405 random_forest_mae:  0.29287713841368584 random_forest mse:  0.1615710730948678 random_forest rmse:  0.40195904405158966
DecisionTreeRegressor mean : 0.3343701399688958 decision tree mae:  0.2986003110419907 decison tree mse:  0.2986003110419907 decison tree rmse:  0.5464433282985444


## Finding and recommedation

from the results of our models, the random forest regression proofs to have the best accuracy since it has the lowest MSE, mean squared error of 0.16 or RMSE of 0.4.


